src/nmod_poly/evaluate_geometric_nmod_vec.c

```diff
@@ -0,0 +1,106 @@
+/*
+    Copyright (C) 2025, Vincent Neiger, Éric Schost
+    Copyright (C) 2025, Mael Hostettler
+
+    This file is part of FLINT.
+
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.  See <https://www.gnu.org/licenses/>.
+*/
+
+#include "nmod.h"
+#include "nmod_poly.h"
+#include "nmod_vec.h"
+
+void
+_nmod_poly_evaluate_geometric_nmod_vec_fast_precomp(nn_ptr vs, nn_srcptr poly, 
+    slong plen, const nmod_geometric_progression_t G, slong len)
+{
+    nn_ptr a, b;
+    slong i, i_min, G_len, a_len;
+
+    G_len = G->len;
+    FLINT_ASSERT(len <= G_len);
+
+    if (plen == 0)
+    {
+        _nmod_vec_zero(vs, G_len);
+        return;
+    }
+
+    for (i_min = 0; i_min < plen; i_min++)
+    {
+        if (poly[i_min] != 0) 
+        {
+            break;
+        }
+    }
+    
+    a_len = plen - i_min;
+    a = _nmod_vec_init(a_len);
+    b = _nmod_vec_init(G->f->length + a_len - 1);
+    
+    for (i = i_min; i < plen; i++)
+    {
+        a[plen - 1 - i] = nmod_mul(G->x[i], poly[i], G->mod);
+    }
+
+    _nmod_poly_mulhigh(b, G->f->coeffs, G->f->length, a, a_len, plen - 1, G->mod);
```

Comment:
``_nmod_poly_mulhigh`` isn't currently optimised; it is even slower than a regular mul. The best way to do a high multiplication currently is to reverse in a temporary buffer and call ``_nmod_poly_mullow``.